### PR TITLE
ExperimentSetMatrix (new version of JointAnalysisMatrix)

### PR DESCRIPTION
**Pull request by utku-ozturk**
This PR addresses two issues:

1. **joint analysis matrix links don't accept multiple values**: (from Trello): Here, I am mapping multiple values of cell_type to one ("H1-hESC (Tier 1)" and "H1-hESC" to "H1-hESC"). And most functionality works. https://data.4dnucleome.org/static-sections/da536eed-248c-4e0b-84f5-f8a390d4add1/. Except the show experiment sets button does not expand search completely. It takes me to [this](https://data.4dnucleome.org/browse/?award.project=4DN&experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29&experiments_in_set.experiment_type.assay_subclass_short=Hi-C&experiments_in_set.experiment_type.display_title=in+situ+Hi-C&experimentset_type=replicate&status=released&status=released+to+project&type=ExperimentSetReplicate). Instead of [this](https://data.4dnucleome.org/browse/?award.project=4DN&experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29&experiments_in_set.biosample.biosource_summary=H1-hESC&experiments_in_set.experiment_type.assay_subclass_short=Hi-C&experiments_in_set.experiment_type.display_title=in+situ+Hi-C&experimentset_type=replicate&status=released&status=released+to+project&type=ExperimentSetReplicate).
(diff is _"&experiments_in_set.biosample.biosource_summary=H1-hESC"_)

**solution:** `JointAnalysisMatrix/VisualBody` component uses `underscore`'s invert function to form URL query from `valueChangedMap` prop.  But underscore's invert requires all of the object's values should be unique and string serializable and this not works if multiple items map to the same value. So new `VisualBody.invert` introduced that allows multiple mappings and converts them to an array.

2. **Softcode number of matrices and their titles**: `JointAnalysisMatrix` allows two matrices: 4DN and ENCODE. Instead user can add additional matrices and customize it.

**solution:** To prevent possible backward compatibility issues, introduced new ExperimentSetMatrix that allows N matrices. sectionKeys prop allows user to determine the matrices and orders. Here is a possible working example:

```
<ExperimentSetMatrix
    key="anyRandomString"  // Required to prevent re-instantiation of component upon window resize & similar.
    session={session}      // Required - hooks in 'session' (boolean) from App.
    sectionKeys={["4DN", "ENCODE", "ENCODE2"]}
    queries={{
            "4DN": {
                "url": "/browse/?experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29&experiments_in_set.biosample.biosource_summary=HFFc6+%28Tier+1%29&experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29+differentiated+to+definitive+endoderm&experimentset_type=replicate&type=ExperimentSetReplicate&award.project=4DN&limit=all",
                "url_fields": [
                       "experiments_in_set.experiment_type.display_title", "lab", "experiments_in_set.biosample.biosource_summary", "status", "lab.display_title",
                       "experiments_in_set.experiment_categorizer.value", "experiments_in_set.experiment_categorizer.field", "experiments_in_set.display_title",
                       "experiments_in_set.accession", "experiments_in_set.experiment_type.assay_subclass_short"
                ]
            },
            "ENCODE": {
                "url": "https://www.encodeproject.org/search/?type=Experiment&biosample_summary=H1&biosample_summary=HFFc6&status!=archived&status!=revoked&limit=all",
                "url_fields": ["assay_slims", "biosample_summary", "assay_term_name", "description", "lab", "status"]
            },
            "ENCODE2": {
                "url": "https://www.encodeproject.org/search/?type=Experiment&biosample_summary=H1-hESC&biosample_summary=HFFc6&status!=archived&status!=revoked&limit=all",
                "url_fields": ["assay_slims", "biosample_summary", "assay_term_name", "description", "lab", "status"]
            }
    }}
    valueChangeMap={{
        "4DN" : {
            "cell_type" : {
                "H1-hESC (Tier 1) differentiated to definitive endoderm" : "H1-DE",
                "H1-hESC (Tier 1)" : "H1-hESC",
                "HFFc6 (Tier 1)" : "HFFc6"
            },
            "state" : {
                "released" : "Submitted",
                "current" : "Submitted",
                "released to project" : "Internal Release",
                "pre-release" : "Internal Release",
                "in review by lab" : "In Submission",
                "in review by project" : "In Submission",
                "submission in progress" : "In Submission",
                "released to lab" : "In Submission",
                "to be uploaded by workflow" : "Planned",
                "planned" : "Planned",
                "archived" : "Out of date",
                "revoked" : "Out of date",
                "deleted" : "Deleted"
            }
        },
        "ENCODE" : {
            "cell_type" : {
                "H1" : "H1-hESC"
            },
            "state" : {
                "released" : "Submitted"
            }
        },
        "ENCODE2" : {
            "cell_type" : {
                "H1" : "H1-hESC"
            },
            "state" : {
                "released" : "Submitted"
            }
        }
    }}
    fieldChangeMap={{
        "4DN"                       : {
            "experiment_category"       : "experiments_in_set.experiment_type.display_title",
            "experiment_type"           : "experiments_in_set.experiment_type.display_title",
            "cell_type"                 : "experiments_in_set.biosample.biosource_summary",
            "sub_cat"                   : "experiments_in_set.experiment_categorizer.value",
            "sub_cat_title"             : "experiments_in_set.experiment_categorizer.field",
            "lab_name"                  : "lab.display_title",
            "short_description"         : "experiments_in_set.display_title",
            "state"                     : "status"
        },
        "ENCODE"                    : {
            "experiment_category"       : "assay_slims",
            "experiment_type"           : "assay_term_name",
            "cell_type"                 : "biosample_summary",
            "lab_name"                  : "lab.title",
            "short_description"         : "description",
            "state"                     : "status"
        },
        "ENCODE2"                    : {
            "experiment_category"       : "assay_slims",
            "experiment_type"           : "assay_term_name",
            "cell_type"                 : "biosample_summary",
            "lab_name"                  : "lab.title",
            "short_description"         : "description",
            "state"                     : "status"
        }
    }}
    groupingProperties={{
        "4DN": ["experiment_type", "sub_cat"],
        "ENCODE": ["experiment_category", "experiment_type"],
        "ENCODE2": ["experiment_category", "experiment_type"]
    }}
    columnGrouping={{
        "4DN": "cell_type",
        "ENCODE": "cell_type",
        "ENCODE2": "cell_type"
    }}
    headerFor={{
        "4DN": (<React.Fragment>
                    <h3 className="mt-2 mb-0 text-300">4DN</h3>
                    <h5 className="mt-0 text-500" style={{ 'marginBottom': -20, 'height': 20, 'position': 'relative', 'zIndex': 10 }}>
                        <a href="/browse/?experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29&experiments_in_set.biosample.biosource_summary=HFFc6+%28Tier+1%29&experiments_in_set.biosample.biosource_summary=H1-hESC+%28Tier+1%29+differentiated+to+definitive+endoderm&experimentset_type=replicate&type=ExperimentSetReplicate&award.project=4DN">Browse all</a> 4DN data-sets
                    </h5>
                </React.Fragment>),
        "ENCODE": (<h3 className="mt-2 mb-0 text-300">ENCODE</h3>),
        "ENCODE2": (<h3 className="mt-2 mb-0 text-300">ENCODE Extra</h3>)
    }}
    sectionStyle={{
        "4DN": {
            "sectionClassName": "col-md-4", //section width
            "rowLabelListingProportion": "balanced" //possible values: wide-label / wide-listing / balanced
        },
        "ENCODE": {
            "sectionClassName": "col-md-4",
            "rowLabelListingProportion": "balanced"
        },
        "ENCODE2": {
            "sectionClassName": "col-md-4",
            "rowLabelListingProportion": "balanced"
        }
    }}
    headerColumnsOrder={["H1-hESC", "H1-DE", "HFFc6"]}
    columnSubGroupingOrder={["Submitted", "In Submission", "Planned", "Not Planned"]}
    titleMap={{
        "_common_name"              : " ",
        "sub_cat"                   : "AnyStringHere",
        "experiment_type"           : "Experiment Type",
        "data_source"               : "Available through",
        "lab_name"                  : "Lab",
        "experiment_category"       : "Category",
        "state"                     : "Submission Status",
        "cell_type"                 : "Cell Type",
        "short_description"         : "Description",
        "award"                     : "Award",
        "accession"                 : "Accession",
        "number_of_experiments"     : "# Experiments in Set",
        "submitted_by"              : "Submitter",
        "experimentset_type"        : "Set Type"
    }}
    fallbackNameForBlankField="None"
/>
```

Output:
https://i.gyazo.com/15ac79d7814f2a4457ebcddec702edbb.png

### CODE REVIEW

**Comment by alexkb0009** [comment]
Glanced through and looks good!
Can we get rid of the old JointAnalysisMatrix and use the new ExperimentSetMatrix in its place?
For any older Static Sections which might use JointAnalysisMatrix, we could probably just "alias" JointAnalysisMatrix in placeholders to refer to ExperimentSetMatrix as well to handle any older Static Sections which might use JointAnalysisMatrix (and if props have changed, could just have JointAnalysisMatrix return/render ExperimentSetMatrix but modify props into the newer/better data structures, like a "transitional" component).
(It'd be easier in long run to maintain one component and not have to do changes 2x+ if want something changed on both JointAnalysisMatrix and ExperimentSetMatrix in future)

**Comment by utku-ozturk** [comment]
@alexkb0009 Yes, you are right: maintaining both components not make sense. So, if it is good to merge:

1. Merge this branch into master
2. Patch existing data: convert all JointAnalysisMatrix to ExperimentSetMatrix
3. Remove JointAnalysisMatrix

File: src/encoded/tests/data/master-inserts/static_section.json [hunk 1]
```diff
@@ -710,7 +710,16 @@
         }
     },
 
-
+    {
+        "name": "experiment-set-data-plans.matrix",
+        "uuid": "d6bf6e5c-e0b4-41c4-bfb5-f56edc480f67",
```
Comment (j1z0) [inline src/encoded/tests/data/master-inserts/static_section.json:715]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js  26
- src/encoded/static/components/static-pages/components/StackedBlockVisual.js  2
         

Clones added
============
- src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js  11
- src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113&bid=18722826)

File: src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js [hunk 2]
```diff
@@ -0,0 +1,600 @@
...
+                    convertedResult[field] = changeMap[convertedResult[field]] || convertedResult[field];
+                }
+            });
+        }
+
+        // Standardized state from status
+        // TODO Use similar by-data-source structure as fieldChangeMap & valueChangeMap
+        if (statusStateTitleMap){
+            const [ stateTitleToSave ] = _.find(_.pairs(statusStateTitleMap), function([titleToSave, validStatuses]){ return validStatuses.indexOf(result.status) > -1; });
+            convertedResult.state = stateTitleToSave || fallbackNameForBlankField;
```
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js:194]: Codacy found an issue: ['statusStateTitleMap' is missing in props validation](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

File: src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+            } else {
+                result[object[keys[i]]] = keys[i];
+            }
+        }
+        return result;
+    }
+
+    constructor(props){
+        super(props);
+        this.blockPopover = this.blockPopover.bind(this);
```
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js:419]: Codacy found an issue: [Use object destructuring.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

File: src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js [hunk 4]
```diff
@@ -373,6 +373,27 @@ class VisualBody extends React.PureComponent {
...
+    /**
+     * replacement of underscore's invert function.
+     * While underscore's invert requires all of object's values should be
+     * unique and string serializable, VisualBody.invert allows multiple
+     * mappings and convert them to array.
+     **/
+    static invert(object) {
+        const result = {};
+        const keys = Object.keys(object);
+        for (var i = 0, length = keys.length; i < length; i++) {
```
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js:385]: Codacy found an issue: [Use object destructuring.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- src/encoded/static/components/static-pages/components/StackedBlockVisual.js  2
- src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js  26
         

Clones added
============
- src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js  11
- src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113&bid=18722826)

File: src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js (hunk 2 above)
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js:194]: Codacy found an issue: ['statusStateTitleMap' is missing in props validation](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

File: src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js (hunk 3 above)
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/ExperimentSetMatrix.js:419]: Codacy found an issue: [Use object destructuring.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

File: src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js (hunk 4 above)
Comment (j1z0) [inline src/encoded/static/components/static-pages/placeholders/JointAnalysisMatrix.js:385]: Codacy found an issue: [Use object destructuring.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5723113)

**Pull request merged**